syntax = "proto3";

import "mars/serialize/protos/value.proto";


message Node1Def {
    string a = 1;
    int32 b1 = 2;
    int32 b2 = 3;
    int32 b3 = 4;
    int64 b4 = 5;
    uint32 c1 = 6;
    uint32 c2 = 7;
    uint32 c3 = 8;
    uint64 c4 = 9;
    float d1 = 10;
    float d2 = 11;
    double d3 = 12;
    bool e = 13;
    Value f = 14;
    Node2Def g = 15;
    repeated Value h = 16;
    repeated Node1Def i = 17;
}


message Node2Def {
    Value a = 1;
    string key = 2;
    repeated int32 data = 3;
    string id = 4;
    string name = 5;
}


message Node3Def {
    oneof value {
        Node1Def n1 = 1;
        Node2Def n2 = 2;
    }
}


message Node4Def {
    map<string, Value> attr = 1;
}